src/instructlab/model/download.py

```diff
@@ -12,20 +16,247 @@
+                    exist_ok=True,
+                )
+            snapshot_download(
+                token=self.hf_token,
+                repo_id=self.repository,
+                revision=self.release,
+                local_dir=os.path.join(self.download_dest, self.repository),
+            )
+        except Exception as exc:
+            click.secho(
+                f"Downloading safetensors model failed with the following HuggingFace Hub error: {exc}",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+
+
+class OCIDownloader(Downloader):
+    """
+    Class to handle downloading safetensors models from OCI Registries
+    We are leveraging OCI v1.1 for this functionality
+    """
+
+    def __init__(self, repository: str, release: str, download_dest: str, ctx) -> None:
+        super().__init__(
+            repository=repository, release=release, download_dest=download_dest
+        )
+        self.repository = repository
+        self.release = release
+        self.download_dest = download_dest
+        self.ctx = ctx
+
+    def _build_oci_model_file_map(self, oci_model_path: str) -> dict:
+        """
+        Helper function to build a mapping between blob files and what they represent
+        """
+        index_hash = ""
+        try:
+            with open(f"{oci_model_path}/index.json", mode="r", encoding="UTF-8") as f:
+                index_ref = json.load(f)
+            match = re.search("sha256:(.*)", index_ref["manifests"][0]["digest"])
+
+            if match:
+                index_hash = match.group(1)
+            else:
+                click.echo(f"could not find hash for index file at: {oci_model_path}")
+                raise click.exceptions.Exit(1)
+        except FileNotFoundError as exc:
+            raise ValueError(f"file not found: {oci_model_path}/index.json") from exc
+        except json.JSONDecodeError as exc:
+            raise ValueError(
+                f"could not read JSON file: {oci_model_path}/index.json"
+            ) from exc
+        except Exception as exc:
+            raise ValueError("unexpected error occurred: {e}") from exc
+
+        try:
+            with open(
+                f"{oci_model_path}/blobs/sha256/{index_hash}",
+                mode="r",
+                encoding="UTF-8",
+            ) as f:
+                index = json.load(f)
+        except FileNotFoundError as exc:
+            raise ValueError(f"file not found: {oci_model_path}/index.json") from exc
+        except json.JSONDecodeError as exc:
+            raise ValueError(
+                f"could not read JSON file: {oci_model_path}/index.json"
+            ) from exc
+        except Exception as exc:
+            raise ValueError("unexpected error occurred: {e}") from exc
+
+        title_ref = "org.opencontainers.image.title"
+        oci_model_file_map = {}
+
+        for layer in index["layers"]:
+            match = re.search("sha256:(.*)", layer["digest"])
```

Comment:
nit: you could have a tiny helper function to extract sha from a layer, then use it here and above instead of implementing it twice.

Reply:
I'm all for helper functions, but whenever I've tried to introduce them in PRs in the past but I've gotten told to just do it in-place in reviews 🤷

Reply:
I don't know the context before, but here you also duplicate the same parsing logic in two places, so maybe it's an additional argument for having a helper doing the job once?

Reply:
yeah
I'll add it again

Reply:
on second thought, i might leave this one as is - it seems like overkill for something that is just 2 lines anyway
I don't know that this is something we will have to do as often as loading a json file for eg to justify a helper function for it

Reply:
added